cairo/ethereum/cancun/trie.cairo

```diff
@@ -372,55 +464,353 @@ func bytes_to_nibble_list{bitwise_ptr: BitwiseBuiltin*}(bytes_: Bytes) -> Bytes
+    tempvar is_nibble_case = nondet %{ memory.get(ids.dict_ptr.key.value.data + ids.level) == ids.nibble %};
+    tempvar is_value_case = nondet %{ int(ids.dict_ptr.key.value.len == ids.level) %};
+
+    static_assert branch == [ap - 4];
+    static_assert dict_ptr == [ap - 3];
+
+    jmp value_case if is_value_case != 0;
+    jmp nibble_case if is_nibble_case != 0;
+    jmp not_nibble_case;
+
+    value_case:
+    let branch = cast([ap - 4], BytesBytesDictAccess*);
+    let dict_ptr = cast([ap - 3], BytesBytesDictAccess*);
+
+    assert dict_ptr.key.value.len = level;
+    assert value = dict_ptr.new_value;
+    assert value_set = 1;
+
+    let dict_ptr_stop = cast([fp + 1], BytesBytesDictAccess*);
+    tempvar stop = (dict_ptr_stop - dict_ptr) - BytesBytesDictAccess.SIZE;
+    tempvar branch = branch;
+    tempvar dict_ptr = dict_ptr + BytesBytesDictAccess.SIZE;
+
+    static_assert branch == [ap - 2];
+    static_assert dict_ptr == [ap - 1];
+    jmp loop if stop != 0;
+    jmp end;
+
+    // Case 1: nibble != key[level], don't include in branch
+    not_nibble_case:
+    let branch = cast([ap - 4], BytesBytesDictAccess*);
+    let dict_ptr = cast([ap - 3], BytesBytesDictAccess*);
+
+    assert_not_zero(dict_ptr.key.value.data[level] - nibble);
+
+    let dict_ptr_stop = cast([fp + 1], BytesBytesDictAccess*);
+    tempvar stop = (dict_ptr_stop - dict_ptr) - BytesBytesDictAccess.SIZE;
+    tempvar branch = branch;
+    tempvar dict_ptr = dict_ptr + BytesBytesDictAccess.SIZE;
+
+    static_assert branch == [ap - 2];
+    static_assert dict_ptr == [ap - 1];
+    jmp loop if stop != 0;
+    jmp end;
+
+    // Case 2: nibble == key[level], include in branch
+    nibble_case:
+    let branch = cast([ap - 4], BytesBytesDictAccess*);
+    let dict_ptr = cast([ap - 3], BytesBytesDictAccess*);
+
+    assert dict_ptr.key.value.data[level] = nibble;
+    assert [branch].key = dict_ptr.key;
+    assert [branch].prev_value = dict_ptr.prev_value;
+    assert [branch].new_value = dict_ptr.new_value;
+
+    let dict_ptr_stop = cast([fp + 1], BytesBytesDictAccess*);
+    tempvar stop = (dict_ptr_stop - dict_ptr) - BytesBytesDictAccess.SIZE;
+    tempvar branch = branch + BytesBytesDictAccess.SIZE;
+    tempvar dict_ptr = dict_ptr + BytesBytesDictAccess.SIZE;
+
+    static_assert branch == [ap - 2];
+    static_assert dict_ptr == [ap - 1];
+    jmp loop if stop != 0;
+    jmp end;
+
+    end:
+    let branche_stop = cast([ap - 2], BytesBytesDictAccess*);
+    let branche_start = cast([fp], BytesBytesDictAccess*);
+    let value = Bytes(cast([fp + 2], BytesStruct*));
+    let value_set = [fp + 3];
+
+    // Fill value_set if it's not set yet. This is just to be able to test against 1
+    // as this would raise if the memory is empty.
+    %{ ids.value_set = memory.get(fp + 3) or 0 %}
+    if (value_set != 1) {
+        let (data: felt*) = alloc();
+        tempvar empty_bytes = Bytes(new BytesStruct(data, 0));
+        assert value = empty_bytes;
+    }
```

Comment:
possible inconsistency here:

value_case => value_set == 1, value = dict_ptr.new_value
but the prover sets: value_set = 0 here

idea: if value_set != 1, `assert cast(value, felt) != 0` ?

Reply:
ah, no ok: there's a speific assigment of `empty_bytes` if the value is not set.
can't we not just assert that the prover did not return the wrong response in the hint?

Reply:
> but the prover sets: value_set = 0 here

this is not possible, the prover cannot override a value. if value_set as been set to 1, then it's 1.

Reply:
I don't understand what you mean by 

> can't we not just assert that the prover did not return the wrong response in the hint?

the idea is that the value is either constraint (value case) or not. I just want to check if it's not constraint, and if not, then put an empty bytes.

It's also possible to do it with no hints and a pointer. I could do

```
local value_set: felt*

// ...

value_case:
assert value_set = alloc();

// ...

if (cast(value_case, felt) == 0) {
    // set empty bytes
}
```

but it's a bit hacky as it uses the fact that you can use a relocatable to have set or not, while we want a bool